### ISSUE TITLE
Firefox 51: This doesn't work

### ISSUE THREAD

**Issue by cusn** [einaregilsson/Redirector#90]
"includePattern": "^https:\\/\\/addons\\.mozilla\\.org\\/.*\\/firefox\\/(.*)",
            "excludePattern": "^https:\\/\\/addons\\.mozilla\\.org\\/zh-CN\\/firefox\\/.*",
            "redirectUrl": "https://addons.mozilla.org/zh-CN/firefox/$1",

**Comment by einaregilsson** [comment]
Ok, will look into it before next release.

**Comment by Gitoffthelawn** [comment]
Interesting.  I was just debugging a somewhat similar rule that suddenly stopped working for the same domain.  Firefox 52.

**Comment by einaregilsson** [comment]
That is weird. Works in Chrome, but I can reproduce this in Firefox. Will look into it soon.

**Comment by Gitoffthelawn** [comment]
Thanks. :)

**Comment by Gitoffthelawn** [comment]
Ping ;)

**Comment by einaregilsson** [comment]
We have another issue for FF52, https://github.com/einaregilsson/Redirector/issues/94, and it turns out there's a bug in that FF version that seriously affects the redirection: https://bugzilla.mozilla.org/show_bug.cgi?id=1345893

So, I'm not gonna spend time on this bug until they've fixed that one, since it might well be related and I'd be wasting my time. I'm pretty sure they'll do a patch version of 52 soon, this bug to me at least seems serious enough for that.

**Comment by Gitoffthelawn** [comment]
Thanks.  Hopefully that will fix this too.  And hopefully they will release a bugfix version before 53.

**Comment by Gitoffthelawn** [comment]
Tested on 52.0.2.  No workie.  But not a big surprise, since Firefox bug 1345893 is still not fixed.

**Comment by einaregilsson** [comment]
Ok. Well, I'll monitor that bug and test again when it's fixed and a patch released.

**Comment by Gitoffthelawn** [comment]
Sounds good.  Thanks.

**Comment by cusn** [comment]
Bug is still not Fixed on Firefox 53.

**Comment by einaregilsson** [comment]
For now we're waiting for the Firefox bug in https://github.com/einaregilsson/Redirector/issues/94 to be fixed, until that's fixed I don't want to spend time working on other bugs, as they might be related to that one.

**Comment by ohohohoo** [comment]
I'm running the recent beta that has fixed the bug described in #94. I can confirm that requesting a.com to be redirected to somewhere.com works great BUT unfortunately the bug described in this issue still exist on the recent beta version of Firefox (Firefox 54).

At first I thought that there was something wrong in my regex rule (I tried to strip src in mozilla addons page), but after several tests I found out that any rule that tried to modify addons.mozilla.org wouldn't work at all, even for a simple thing like (using wildcard not regex) *mozilla.org* to be redirected to google.com.

Opening mozilla.org (without addons subdomain) would get redirected automatically to google but not addons.mozilla.org.

I looked at the error console and found something that basically says Redirector extension blocked by CSP.  (note that the error message doesn't always appear).

"Content Security Policy: The page’s settings blocked the loading of a resource at self (“script-src moz-extension://b71bc8c2-cc9b-443f-bf6c-839bc7875757”). Source: call to eval() or related function blocked by CSP."

I looked at the mozilla webextension page regarding webext and csp blocking and found that I need to add CSP rule and so I added it manuallly by editing the extension and add this:

"content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self';"

But still the problem persist and I'm lost now. Any help would be great :)

**Comment by einaregilsson** [comment]
Ok, very interesting! So the background script is blocked by CSP! Not sure how to fix that, but it's definitely something I'll have to look into for future versions of Redirector, especially if more sites start using CSP. Thanks for letting me know about this!

**Comment by Gitoffthelawn** [comment]
It does seem that more and more sites are using CSP.  I wonder if this bug can be fixed independent of the one described in #94, as there is a reliable test site: addons.mozilla.org that shouldn't be impacted by #94.

**Comment by einaregilsson** [comment]
Maybe it could. I won't be able to look at it for the next couple of weeks at least though, pretty busy right now with my actual work :)

**Comment by ohohohoo** [comment]
I just found out (regarding the addons.mozilla.org unaffected by any redirector rules) apparently it doesn't have anything to do with CSP (perhaps). I disabled CSP completely in Firefox (also just found out that CSP can be disabled in Firefox while I was checking out the about:config) and then make sure that CSP is ignored by simply executing `javascript:alert(1)` while on addons.mozilla.org but still the problem persist and so I think the problem lies elsewhere :(

**Comment by Gitoffthelawn** [comment]
@einaregilsson No problem. Definitely keep food on the table!

**Comment by Gitoffthelawn** [comment]
I think I figured out the issue regarding `addons.mozilla.org`.  Firefox now considers this to be a "special" host, and therefore I think WebExtensions can't touch it.

I understand their rationale, but a trusted extension like Redirector would be so useful when applied to that host.

**Comment by einaregilsson** [comment]
Ok. Well, that makes sense I guess. Do you know about any blog post or anything where they talk about it?

**Comment by Gitoffthelawn** [comment]
I remember reading about it, but I don't have the reference on hand.  I've tested quite a few WebExtension API extensions, and none of them can touch `addons.mozilla.org`.  It's really unfortunate, but IIRC they are doing this to try to protect systems from attacks that modify connections to that host or the appearance of pages from that host.  Of course, if a user has a bad extension installed, the system is already compromised.

**Comment by einaregilsson** [comment]
Ok, well, I guess it makes sense. Closing this issue then.

**Comment by Gitoffthelawn** [comment]
Einar, if you have a moment, could you post a bug on bugzilla linking here to let Mozilla know about the unintended side-effects of their new implementation?

I really miss Redirector working on AMO.

**Comment by einaregilsson** [comment]
I don't have an account there, and it seems like a hassle. Anyone else is free to do it, but I really don't think they're going to change this, and I understand where they're coming from.

**Comment by Gitoffthelawn** [comment]
> ...but I really don't think they're going to change this...

I'm usually of the philosophy that it never hurts to ask, but I've noticed that the people at Mozilla seem to be less and less interested in actual user feedback.  It's a big turn around from the Mozilla of old.

That said, I too understand the rationale, although I think carefully inspected and approved extensions could be exceptions.

**Comment by Gitoffthelawn** [comment]
I found what looks like the official bugzilla bug regarding this issue:
https://bugzilla.mozilla.org/show_bug.cgi?id=1310082

**Comment by einaregilsson** [comment]
Well, if they ever allow this again it'll work, but for now this issue remains closed.

**Comment by Gitoffthelawn** [comment]
Yep.  I concur.
